src/gridcoin/staking/spam.h

```diff
@@ -0,0 +1,218 @@
+            return false;
+        }
+
+        const int64_t stake_time = MaskStakeTime(coinstake.nTime);
+        const auto& timestamps = iter_pair->second;
+        const auto end = timestamps.end();
+
+        return std::find(timestamps.begin(), end, stake_time) != end;
+    }
+
+    //!
+    //! \brief Remove a kernel proof for an orphan block.
+    //!
+    //! \param coinstake Coinstake transaction from the block.
+    //!
+    void ForgetOrphan(const CTransaction& coinstake)
+    {
+        const COutPoint& stake_prevout = coinstake.vin[0].prevout;
+        auto iter_pair = m_orphan_proofs_seen.find(stake_prevout);
+
+        if (iter_pair == m_orphan_proofs_seen.end()) {
+            return;
+        }
+
+        const int64_t stake_time = MaskStakeTime(coinstake.nTime);
+        auto& timestamps = iter_pair->second;
+        auto end = timestamps.end();
+
+        timestamps.erase(std::remove(timestamps.begin(), end, stake_time), end);
+
+        if (timestamps.empty()) {
+            m_orphan_proofs_seen.erase(iter_pair);
+        }
+    }
+
+    //!
+    //! \brief Fill the container with proof hashes near the chain tip.
+    //!
+    //! \param pindex Points to the block index for the chain tip.
+    //!
+    void Refill(const CBlockIndex* pindex)
+    {
+        for (size_t i = 0; pindex && i < m_proofs_seen.size(); ++i) {
+            pindex = pindex->pprev;
+        }
+
+        for (; pindex; pindex = pindex->pnext) {
+            Remember(pindex->hashProof);
+        }
+    }
+
+private:
+    //!
+    //! \brief A hash table of recently-observed kernel proof hashes.
+    //!
+    //! The size of the container must be a power of 2. Realistically, the 2048
+    //! entries are likely overkill since a block must:
+    //!
+    //!  - satisfy the requirements of the staking protocol
+    //!  - contain a valid coinstake transaction (previously unspent input)
+    //!  - produce a proof not already in this collection
+    //!
+    //! The extra wiggle room gives the container more tolerance for collisions
+    //! by increasing the output range of the hash function. The hash table can
+    //! store about two days' volume of staking proofs.
+    //!
+    std::array<uint256, 2048> m_proofs_seen;
+
+    //!
+    //! \brief A set of observed stake inputs from orphan blocks.
+    //!
+    //! Orphan blocks do not undergo proof-of-stake validation, so we cannot
+    //! store the computed proof hash. This collection contains outpoints of
+    //! staked inputs mapped to timestamps of the coinstake transactions.
+    //!
+    std::map<COutPoint, std::vector<int64_t>> m_orphan_proofs_seen;
+
+    //!
+    //! \brief Get the hash table slot offset for a proof hash.
+    //!
+    //! \param Proof hash of the stake used to generate a block.
+    //!
+    //! \return Hash of the proof hash in the range of the size of the proofs
+    //! table.
+    //!
+    size_t GetOffset(const uint256& hashProof) const
+    {
```

Comment:
What does this replace in the current code?

Reply:
Ah... nothing. I think you are implementing something new here required by the binnning (hash bin) scheme you are implementing...

Reply:
That's right&mdash;this function picks the bin for a proof hash since the hash table has a fixed size. Since the original implementation never removed entries from `setStakeSeen`, I figured we didn't need the overhead of one of the `std` maps. New entries just overwrite the earlier entries. This just guards the chain tip, so there's no need to remember every proof from the beginning of time.

Reply:
I like it.